widgetssdk/src/main/java/com/glia/widgets/chat/controller/ChatController.kt

```diff
@@ -1262,13 +1262,17 @@ class ChatController(
     ): List<ChatMessageInternal>? {
         return if (newHistory.isNullOrEmpty() || oldHistory.isNullOrEmpty()) {
             newHistory
-        } else newHistory.filter { isNewMessage(it.chatMessage) }
+        } else newHistory.filter { isNewMessage(oldHistory, it.chatMessage) }
     }
 
-    private val oldMessageIdsHash: MutableSet<String> by lazy { mutableSetOf() }
-
-    private fun isNewMessage(newMessage: ChatMessage): Boolean =
-        oldMessageIdsHash.add(newMessage.id)
+    private fun isNewMessage(oldHistory: List<ChatItem>?, newMessage: ChatMessage): Boolean =
```

Comment:
Just interesting, what was the cause of the message duplication with the previous solution?

Reply:
In the method `removeDuplicates` we had:
```
        return if (newHistory.isNullOrEmpty() || oldHistory.isNullOrEmpty()) {
            newHistory
        } else
```
So, in the beginning, it added all messages without adding their ids to the `oldMessageIdsHash`.

Reply:
May it be productive to add all ids to hash at this point instead of iterating through all old messages for every new one?

Reply:
I believe it does not make a lot of sense if we talking about not huge list. Instead, we will not have inconsistency between two collections.